### PR TITLE
Scotty Connector for Apache Storm

### PR DESCRIPTION
**Pull request by tuterbatuhan**
A connector implementation for Apache Storm including window function implementation and a working demo

### CODE REVIEW

**Comment by PhilippGrulich** [comment]
Hey @tuterbatuhan,

looks really good.
I added some minor comments.
I'm not sure if we should also put the storm native aggregation functions in the connector. But they are just used in the demo right? I think then its ok.

If you fixed the comments. I will merge the pull request.

Best,
Philipp

File: storm-connector/pom.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        <jdk.version>1.8</jdk.version>
+        <maven-compiler-plugin.version>3.7.0</maven-compiler-plugin.version>
+        <maven-shade-plugin.version>3.1.0</maven-shade-plugin.version>
+        <scotty.version>0.4</scotty.version>
+        <storm.version>1.2.2</storm.version>
+        <args4j.version>2.33</args4j.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/pom.xml]: I assume you mean the CSV dependency. That is deleted from the pom.

File: storm-connector/pom.xml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.apache.storm/storm-core -->
+        <dependency>
+            <groupId>org.apache.storm</groupId>
+            <artifactId>storm-core</artifactId>
+            <version>${storm.version}</version>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/args4j/args4j -->
+        <dependency>
+            <groupId>args4j</groupId>
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/pom.xml]: I assume you mean the args4j and scala library. They are deleted from the pom.

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/ScottyWindowTopology.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+/*
+ * @author Batuhan Tüter
+ * Runner class for Scotty on Storm
+ *
+ *
+ * Input data should be in the following format:
+ * <key,value,timeStamp>, timeStamp is type Long
+ * */
+
+public class ScottyWindowTopology {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/ScottyWindowTopology.java]: Done

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MinBolt.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.storm.topology.base.BaseWindowedBolt;
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.windowing.TupleWindow;
+
+import java.util.List;
+import java.util.Map;
+
+public class MinBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MinBolt.java]: Deleted

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MeanBolt.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.windowing.TupleWindow;
+
+import java.util.List;
+import java.util.Map;
+
+
+public class MeanBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MeanBolt.java]: Deleted

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MaxBolt.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.storm.topology.base.BaseWindowedBolt;
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.windowing.TupleWindow;
+
+import java.util.List;
+import java.util.Map;
+
+public class MaxBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/MaxBolt.java]: Deleted

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/QuantileBolt.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.windowing.TupleWindow;
+
+import java.util.List;
+import java.util.Map;
+
+
+public class QuantileBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/QuantileBolt.java]: Deleted

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumSlidingWindowBolt.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
+
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements.  See the NOTICE file distributed with
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumSlidingWindowBolt.java]: The whole implementation is deleted from the PR

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumSlidingWindowBolt.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Computes sliding window sum
+ */
+public class SumSlidingWindowBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumSlidingWindowBolt.java]: Deleted

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumTumblingWindowBolt.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/*
+ * Computes tumbling window average
+ */
+public class SumTumblingWindowBolt extends BaseWindowedBolt {
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/storm/SumTumblingWindowBolt.java]: Deleted

File: storm-connector/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/pom.xml]: Args4J dependency is removed. Scala tuples are used in the Mean window function

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/KeyedScottyWindowOperator.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer outputFieldsDeclarer) {
+        outputFieldsDeclarer.declare(new Fields("key", "sum"));
+    }
+
+    @Override
+    public void execute(Tuple tuple, BasicOutputCollector basicOutputCollector) {
+        //The input is a tuple
+        Key currentKey = (Key) tuple.getValue(0);
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/KeyedScottyWindowOperator.java]: We don't assume that the input tuple has 3 fields. It is defined by the source. In our demo case, it is defined in the DataGeneratorSpout, line 56.

File: storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/scotty/MeanScotty.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
+package de.tub.dima.scotty.stormconnector.demo.windowFunctions.scotty;
+
+import de.tub.dima.scotty.core.windowFunction.AggregateFunction;
+import scala.Tuple2;
```
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/src/main/java/de/tub/dima/scotty/stormconnector/demo/windowFunctions/scotty/MeanScotty.java]: Scala Tuples are no longer used

File: storm-connector/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tuterbatuhan) [reply storm-connector/pom.xml]: Scala Tuples are also removed from the dependencies

**Pull request merged**